File: opaia_backend/app/crud/rent.py
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_
from datetime import date
from app.models.rent import Rent
from app.schemas.rent import RentCreate
from fastapi import HTTPException

def verificar_disponibilidade(db: Session, car_id: int, data_inicio: date, data_fim: date):
    conflitos = db.query(Rent).filter(
        Rent.car_id == car_id,
        or_(
            and_(Rent.start_date <= data_inicio, Rent.end_date >= data_inicio),
            and_(Rent.start_date <= data_fim, Rent.end_date >= data_fim),
            and_(Rent.start_date >= data_inicio, Rent.end_date <= data_fim),
        )
    ).first()

    if conflitos:
        raise HTTPException(status_code=400, detail="Carro indisponível para o período selecionado.")

def criar_aluguel(db: Session, aluguel: RentCreate, user_id: int):
    verificar_disponibilidade(db, aluguel.car_id, aluguel.data_inicio, aluguel.data_fim)

    novo_aluguel = Rent(
        user_id=user_id,
        car_id=aluguel.car_id,
        start_date=aluguel.data_inicio,
        end_date=aluguel.data_fim,
        status="pendente"
    )

    db.add(novo_aluguel)
    db.commit()
    db.refresh(novo_aluguel)
    return novo_aluguel

File: opaia_motors_frontend/lib/models/rent.dart
class Rent {
  final int id;
  final int userId;
  final int carId;
  final String startDate;
  final String endDate;
  final String status;

  Rent({
    required this.id,
    required this.userId,
    required this.carId,
    required this.startDate,
    required this.endDate,
    required this.status,
  });

  factory Rent.fromJson(Map<String, dynamic> json) {
    return Rent(
      id: json['id'],
      userId: json['user_id'],
      carId: json['car_id'],
      startDate: json['start_date'],
      endDate: json['end_date'],
      status: json['status'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'car_id': carId,
      'start_date': startDate,
      'end_date': endDate,
      'status': status,
    };
  }
}
